cpp/src/bitmask/bit_container.cuh

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) 2018, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef _BIT_CONTAINER_H_
+#define _BIT_CONTAINER_H_
+
+#pragma once
+
+#include <string.h>
+
+#include "cudf.h"
+#include "bit_container.h"
+#include "rmm/rmm.h"
+#include "utilities/error_utils.h"
+#include <cuda_runtime_api.h>
+
+
+/* ---------------------------------------------------------------------------- */
+/**
+ * @Synopsis  Base class for managing bit masks.
+ */
+/* ---------------------------------------------------------------------------- */
+class Util {
+public:
+  __device__ __host__
+  Util(bit_container_t *valid, int bitlength): valid_(valid), bitlength_(bitlength) {}
+
+  __device__ __host__
+  Util(const Util &util): valid_(util.valid_), bitlength_(util.bitlength_) {}
+
+  /**
+   * Check to see if a record is Valid (aka not null)
+   *
+   * @param[in] record_idx   the record index to check
+   *
+   * @return  true if record is valid, false if record is null
+   */
+  template <typename T>
+  __host__ __device__
+  bool IsValid(T record_idx) const {
+    return is_valid(valid_, record_idx);
+  }
+
+  /**
+   * Set a bit
+   *
+   * @param[in] record_idx    the record index
+   */
+  template <typename T>
+  __host__ __device__
+  void SetBit(T record_idx) {
+    int rec = which_word(record_idx);
+    int bit = which_bit(record_idx);
+
+    valid_[rec] = valid_[rec] | (1U << bit);
```

Comment:
If two concurrent CUDA threads (say, threads in the same warp) call this function on nearby indices (e.g. based on `threadIdx.x`), you have a race condition. So perhaps this should be implemented with an atomic in CUDA. You may need different implementations for host and device code. Alternatively, you could make an `atomicSetBit` if you expect to need atomic and non-atomic versions.